forge/db/controllers/Device.js

```diff
@@ -271,5 +271,176 @@ module.exports = {
+        }
+
+        // Prepare the updates
+        const logEntries = []
+        const devicesToUpdate = []
+        const projectSettings = await assignToProject?.getSetting('deviceSettings') || {}
+        const projectTargetSnapshotId = projectSettings.targetSnapshot || null
+        for (let index = 0; index < devices.length; index++) {
+            const device = devices[index]
+            if (!assignTo) {
+                // ### Remove device from application/project ###
+                let previousOwner
+                if (!device.Project && device.Application) {
+                    previousOwner = {
+                        id: device.Application.id,
+                        name: device.Application.name,
+                        isApplicationOwned: true
+                    }
+                } else if (device.Project) {
+                    previousOwner = {
+                        id: device.Project.id,
+                        name: device.Project.name,
+                        isApplicationOwned: false
+                    }
+                } else {
+                    continue // Device is already unassigned - nothing to do
+                }
+                device.ProjectId = null // unassign from project
+                device.ApplicationId = null // unassign from application
+                device.targetSnapshotId = null // clear the target snapshot
+                device.DeviceGroupId = null // clear the deviceGroup
+                // RE: disable developer mode - this behaviour is aligned with the device update API endpoint
+                device.mode = 'autonomous' // disable developer mode
+                devicesToUpdate.push(device)
+                logEntries.push({
+                    logger: app.auditLog.Team.team.device.unassigned,
+                    params: [user, null, team, previousOwner, device]
+                })
+                if (previousOwner.isApplicationOwned) {
+                    logEntries.push({
+                        logger: app.auditLog.Application.application.device.unassigned,
+                        params: [user, null, previousOwner, device]
+                    })
+                } else {
+                    logEntries.push({
+                        logger: app.auditLog.Project.project.device.unassigned,
+                        params: [user, null, previousOwner, device]
+                    })
+                }
+            } else if (assignTo === 'instance') {
+                // ### Add device to instance ###
+                if (device.ownerType === 'instance' && device.Project?.id === assignToProject.id) {
+                    // Device is already assigned to this instance - nothing to do
+                    continue
+                } else {
+                    device.ProjectId = assignToProject.id
+                    device.ApplicationId = assignToProject.ApplicationId
```

Comment:
When assigning to an instance, `ApplicationId` should be null. My understanding is ProjectId and ApplicationId should never both be set.

Reply:
Hmmmm. very well spotted.

I believe, since the target is being set to the value of an existing/valid source `project` the value is inconsequential and the tests (indirectly) confirm this: https://github.com/FlowFuse/flowfuse/pull/4336/files/05cc75601c14f978b6d5d068c25768c135063863#diff-23b65c9650241a9666db727191beff98fe97e80a3a0175102bebd291692ca6abR709 but for correctness here and now, this should probably be `null`ed

I will switch branch and double check.

ta.

Reply:
checked this over. we do need to set this null.

resolved in [f070a56](https://github.com/FlowFuse/flowfuse/pull/4336/commits/f070a56a4f81f17edf4ca6eb1f906be96d411a11)

Thanks.

Reply:
Conceptually, I was always in favour of setting applicationId when assigned to an instance in that application - as it makes retrieving a list of all devices within the application trivial. But that wasn't the path taken, so yes, we now need to ensure only one or the other is set.